lighthouse-core/computed/unused-javascript-summary.js

```diff
@@ -0,0 +1,173 @@
+ */
+'use strict';
+
+const makeComputedArtifact = require('./computed-artifact.js');
+// const ByteEfficiencyAudit = require('../audits/byte-efficiency/byte-efficiency-audit.js');
+
+/**
+ * @typedef WasteData
+ * @property {Uint8Array} unusedByIndex
+ * @property {number} unusedLength
+ * @property {number} contentLength
+ */
+
+/**
+ * @typedef ComputeInput
+ * @property {string} url
+ * @property {LH.Crdp.Profiler.ScriptCoverage[]} scriptCoverages
+ * @property {LH.Artifacts.Bundle=} bundle
+ */
+
+/**
+ * @typedef Summary
+ * @property {string} url
+ * @property {number} wastedBytes
+ * @property {number} totalBytes
+ * @property {number} wastedBytes
+ * @property {number=} wastedPercent
+ * @property {Record<string, number>=} sourcesWastedBytes
+ */
+
+class UnusedJavascriptSummary {
+  /**
+   * @param {LH.Crdp.Profiler.ScriptCoverage} scriptCoverage
+   * @return {WasteData}
+   */
+  static computeWaste(scriptCoverage) {
+    let maximumEndOffset = 0;
+    for (const func of scriptCoverage.functions) {
+      maximumEndOffset = Math.max(maximumEndOffset, ...func.ranges.map(r => r.endOffset));
+    }
+
+    // We only care about unused ranges of the script, so we can ignore all the nesting and safely
+    // assume that if a range is unexecuted, all nested ranges within it will also be unexecuted.
+    const unusedByIndex = new Uint8Array(maximumEndOffset);
+    for (const func of scriptCoverage.functions) {
+      for (const range of func.ranges) {
+        if (range.count === 0) {
+          for (let i = range.startOffset; i < range.endOffset; i++) {
+            unusedByIndex[i] = 1;
+          }
+        }
+      }
+    }
+
+    let unused = 0;
+    for (const x of unusedByIndex) {
+      unused += x;
+    }
+
+    return {
+      unusedByIndex,
+      unusedLength: unused,
+      contentLength: maximumEndOffset,
+    };
+  }
+
+  /**
+   * @param {string} url
+   * @param {ReturnType<typeof UnusedJavascriptSummary.determineLengths>} lengths
+   * @return {Summary}
+   */
+  static createItem(url, lengths) {
+    const wastedRatio = (lengths.unused / lengths.content) || 0;
+    const wastedBytes = Math.round(lengths.content * wastedRatio);
+
+    return {
+      url: url,
+      totalBytes: lengths.content,
+      wastedBytes,
+      wastedPercent: 100 * wastedRatio,
+    };
+  }
+
+  /**
+   * @param {WasteData[]} wasteData
+   */
+  static determineLengths(wasteData) {
+    let unused = 0;
+    let content = 0;
+    // TODO: this is right for multiple script tags in an HTML document,
+    // but may be wrong for multiple frames using the same script resource.
+    for (const usage of wasteData) {
+      unused += usage.unusedLength;
+      content += usage.contentLength;
+    }
+
+    return {
+      content,
+      unused,
+    };
+  }
+
+  /**
+   * @param {WasteData[]} wasteData
+   * @param {LH.Artifacts.Bundle} bundle
+   */
+  static createSourceWastedBytes(wasteData, bundle) {
```

Comment:
were any of these functions changed or just moved?

Reply:
mostly moved. `createSourceWastedBytes` prob has the biggest change. was called `createBundleMultiData` and calculated the individual source unused sizes as transfer size. Now it's split bw/n the computed artifact (determines all source file unused sizes as response size) and the audit (grabs the top few and multiplies by transfer ratio)

luckily the tests guarantee no breaking changes :)

Reply:
cool yeah I caught the split 👍